<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>Document</title>
  </head>
  <body>
    <h1>首先整理我们常见并需要处理的错误类型</h1>
    <p>js错误</p>
    <p>网络错误</p>
    <p>资源错误</p>
    <hr />
    <h1>错误收集</h1>
    <p>得到需要我们关注的错误点后，我们要如何捕获错误</p>
    <hr />
    <h2>try/catch</h2>
    <i>能捕获常规运行时错误，语法错误和异步错误不行</i>
    <hr />
    <h2>window.onerror</h2>
    <i>
      这是一个全局的错误收集方法，当 JS 运行时错误发生时，window 会触发一个
      ErrorEvent 接口的 error 事件。
    </i>
    <p>
      如果你定义了onerror，但是使用vue时，没有启用Vue.config.errorHandler，那么有很多异常都抓不到（不能捕获语法错误和资源错误）。Vue希望你要定义它，否则异常不会抛出去的。
    </p>
    <code>
      window.onerror = function(message, source, lineno, colno, error) { ... }
    </code>
    <p>参数：</p>
    <pre>
          message：错误信息（字符串）。可用于HTML onerror=""处理程序中的event。
          source：发生错误的脚本URL（字符串）
          lineno：发生错误的行号（数字）
          colno：发生错误的列号（数字）
          error：Error对象（对象）
    </pre>
    <hr />
    <h2>window.addEventListener('error')</h2>
    <i>
      当一项资源（如图片或脚本）加载失败，加载资源的元素会触发一个 Event 接口的
      error 事件，这些 error 事件不会向上冒泡到 window，但能被捕获
    </i>
    <hr />
    <h2>window.addEventListener("unhandledrejection");</h2>
    <i>此方法可以对promise的错误做全局统一处理</i>
    <hr />
    <h2>errorHandler（vue）</h2>
    <p>Vue中最广泛使用的异常处理方式，设置在全局</p>
    <code>
      Vue.config.errorHandler = function(error, vm, info) { ... }
    </code>
    <p>参数：</p>
    <pre>
        vm: vue本身
        error：Error对象
        info：错误信息
    </pre>
    <hr />
    <h2>warnHandler（vue）</h2>
    <p>用来捕获Vue warning。记住在生产环境是不起作用的</p>
    <code>
      Vue.config.warnHandler = function(msg, vm, trace) { ... }
    </code>
    <p>参数：</p>
    <pre>
        msg: 错误信息
        vm: vue本身
        trace:是组件的继承关系追踪
    </pre>
    <hr />
    <h2>errorCaptured（vue）</h2>
    <p>属于生命周期的一部分</p>
    <p>当捕获一个来自子孙组件的错误时被调用</p>
    <p>
      我们可以通过mixin将所有组件混入一个errorCaptured方法，任何组件的错误都将被抛出
    </p>
    <code>
      errorCaptured(err: Error, vm: Component, info: string){ ... }
    </code>
    <p>参数：</p>
    <pre>
        vm: 目标组件
        error：Error对象
        info：错误信息
    </pre>
    <p>
      问题：需要考虑的是，errorCaptured 与
      Vue.config.errorHandler是否会有功能上的重叠，导致一个错误上报多次
    </p>
    <p>解决：此钩子可以返回 false 以阻止该错误继续向上传播。</p>
    <hr />
    <h2>script跨域（通常不太可能发生在线上）</h2>
    <p>一般由后端配置Access-Control-Allow-Origin、前端script加crossorigin</p>
    <p>
      若非要由系统抛出错误，我们可以封装一个方法来改写被使用的方法，使其可以被try/catch
    </p>
    <pre>
        function wrapErrors(fn) {
          if (!fn.__wrapped__) {
            fn.__wrapped__ = function () {
              try {
                return fn.apply(this, arguments);
              } catch (e) {
                throw e;
              }
            };
          }
          return fn.__wrapped__;
        }
        wrapErrors(foo)()
    </pre>
    <p>通过try/catch抛出错误，能被window.onerror捕获</p>
    <p>否则会直接抛出script error</p>
  </body>
</html>
<style>
  code {
    padding: 12px;
    background-color: #eee;
  }
</style>
